configure.py

```diff
@@ -2402,5 +2403,56 @@ def create_building_system(args):
     generate_compdb('compile_commands.json', ninja, args.buildfile, selected_modes)
 
 
+def configure_for_mode(mode, args):
+    # TODO: use Ninja multi-config
+    source_dir = os.path.realpath(os.path.dirname(__file__))
+    build_dir = os.path.join(source_dir, 'build', mode)
+
+    settings = {
+        'CMAKE_BUILD_TYPE': mode,
+        'CMAKE_CXX_COMPILER': args.cxx,
+        'CMAKE_C_COMPILER': args.cc,
+        'CMAKE_C_FLAGS': args.user_cflags,
+        'CMAKE_EXE_LINKER_FLAGS': semicolon_separated(args.user_ldflags),
+        'CMAKE_EXPORT_COMPILE_COMMANDS': 'ON',
+        'CMAKE_EXPORT_NO_PACKAGE_REGISTRY': 'ON',
+    }
+    if args.date_stamp:
+        settings['Scylla_DATE_STAMP'] = args.date_stamp
+
+    cmake_command = ['cmake']
+    cmake_command += [f'-D{var}={value}' for var, value in settings.items()]
+    cmake_command += ['-G', 'Ninja',
+                      '-B', build_dir,
+                      '-S', source_dir]
+    subprocess.check_call(cmake_command, shell=False, cwd=source_dir)
+    return build_dir, os.path.join(build_dir, 'build.ninja')
+
+
+def configure_with_cmake(args):
+    ninja = find_ninja()
+    with open('build.ninja', 'w', encoding='utf-8') as main_build_file:
+        ninja_opts = ''
+        if args.verbose:
+            ninja_opts += ' --verbose'
+        main_build_file.write(textwrap.dedent(f'''
+          rule ninja
+            command = {ninja} {ninja_opts} -C $dir $target
+            restat = 1
```

Comment:
Should be in separate pools, we don't want dev-build and release-build to run in parallel. Unfortunately it will reduce performance since we'll have more jobs running with reduced concurrency.

Can we make cmake generate ninja files that can be included together? that is, ninja files that specify target paths relative to the source directory and not the target directory? Then we could build everything without recursive invocations.

Reply:
> Should be in separate pools, we don't want dev-build and release-build to run in parallel. Unfortunately it will reduce performance since we'll have more jobs running with reduced concurrency.

i see.

> 
> Can we make cmake generate ninja files that can be included together? that is, ninja files that specify target paths relative to the source directory and not the target directory? Then we could build everything without recursive invocations.

with the help of mult-config generator, we can generate multiple build files for each build mode, and CMake creates `build-${mode}.ninja` as wrappers to build the mode specific targets. the wrapper rule looks like
```ninja
include CMakeFiles/impl-Debug.ninja

build test: phony test$:Debug
build seastar/test: phony seastar/test$:Debug
```
and the top-level `build.ninja` just creates alias for the default mode, in this question, `Debug`, like
```
include CMakeFiles/impl-Debug.ninja
...
build tracing/all: phony tracing/all$:Debug
```

technically, we can also use a grand wrapper which includes all the available per-mode build files. but i think it's kind of hackish, as we'd be exploring the internal implantation of CMake.